Ubuntu/History/historymanager.cpp

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (C) 2021 UBports Foundation
+ *
+ * Authors:
+ *  Lionel Duboeuf <lduboeuf@ouvaton.org>
+ *
+ * This file is part of history-service.
+ *
+ * history-service is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * history-service is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "historymanager.h"
+#include "eventview.h"
+#include "manager.h"
+#include <QDebug>
+#include <QJSValueList>
+#include <QTimer>
+
+
+HistoryManager::HistoryManager(QObject *parent) :
+    QObject(parent), mDeletedEventsCount(0), mEventsCount(0), mPendingOperation(OperationError::NO_ERROR)
+{
+}
+
+void HistoryManager::removeEvents(int eventType, const QString &from, const QJSValue &callback)
+{
+    if (!callback.isCallable()) {
+        qCritical() << "no callback found!";
+        return;
+    }
+
+    QJSValue result(callback);
+    if (mPendingOperation) {
+        result.call(QJSValueList { 0, 0, OperationError::OPERATION_ALREADY_PENDING });
+        qWarning() << "there is a pending operation, request cancelled";
+        return;
+    }
+
+    QDateTime fromDate = QDateTime::fromString(from, Qt::ISODate);
+    History::EventType type = (History::EventType) eventType;
+
+    if (type == History::EventTypeNull|| !fromDate.isValid()) {
+        result.call(QJSValueList { 0, 0, OperationError::OPERATION_INVALID });
+        qWarning() << "invalid type or date, request cancelled";
+        return;
+    }
+
+    History::Filter queryFilter(History::FieldTimestamp, QVariant(from), History::MatchLess);
+    History::Sort querySort(History::FieldTimestamp, Qt::DescendingOrder);
+
+    if (!queryFilter.isValid()) {
+        result.call(QJSValueList { 0, 0, OperationError::OPERATION_INVALID });
+        qWarning() << "invalid filter, operation cancelled";
+        return;
+    }
+
+    if (!mView.isNull()) {
+        mView->disconnect(this);
+    }
+
+    mEventsCount = History::Manager::instance()->getEventsCount(type, queryFilter);
+    mDeletedEventsCount = 0;
+    result.call(QJSValueList { mEventsCount, mDeletedEventsCount, OperationError::NO_ERROR });
+
+    if (mEventsCount == 0) {
+        return;
+    }
+
+    mPendingOperation = true;
+    mView = History::Manager::instance()->queryEvents(type, querySort, queryFilter);
+
+    connect(mView.data(), &History::EventView::eventsRemoved,
```

Comment:
So, I had a better look at the existing code, and my understanding is that the code can be such simpler than this (but please let me know, if I'm mistaken):

1. I suggest changing the `HistoryDaemon::removeEvents()` method to run a transaction (like the other removeEvents does), so that it either succeeds (= deletes all the events) or fails (= deletes nothing)
2. Add a `RemovedCount` *output* parameter to `RemoveEventsBy`
3. Change your `ManagerDBus::removeEvents()` to return the `QDBusPendingCall` (that is the result of the `mInterface->asyncCall()`)
4. in the Manager class your new `removeEvents()` method should also take a callback (you could use `std::function<void(int,bool)>`, that is a callback that will have a int parameter — that's the result of the `RemoveEvents` D-Bus call — and a boolean — that indicates if an error occurred). Use `QDBusPendingCallWatcher` to decide when to invoke the callback.
5. No need to listen for any signals here, or to create a view. In this point, you'll have a call similar to this:
```cpp
auto onCompleted = [this, callback](int removedCount, bool isError) {
        QJSValue result(callback);
        result.call(QJSValueList { removed, isError });
}
EventViewPtr view = History::Manager::instance()->removeEvents(type, querySort, queryFilter, onCompleted);
```

Of course, many variants are possible. If you think that a boolean is not enough and you really want to have an error code, you can make it an `int`.

Reply:
If you need some examples for QDBusPendingCallWatcher, you can have a look at https://github.com/ubports/media-hub/blob/xenial_-_mediahubqt/src/lib/MediaHub/player.cpp#L373-L389

Reply:
> So, I had a better look at the existing code, and my understanding is that the code can be such simpler than this (but please let me know, if I'm mistaken):

What about the signal &History::EventView::eventsRemoved which is triggered in each batch ? ( e.g ) when deleting 1000 records, it will signal every 200 deleted records and user should be notified each time. Will your solution work ?

Reply:
ah i see, it is for having a "finihed" signal not for having batch notifocation.
So still need to keep the eventdRemoved connect and invalidated ?

Reply:
ah well, i've noticed that the eventsRemoved is sometimes not triggered after each batch operation , probably due to the loop, so i think i might consider your solution :)